### ISSUE TITLE
Are two way mappings always required?  Will fields in the data that are not configured in the mapping cause errors? 

### ISSUE THREAD

**Issue by jbeckton** [nartc/mapper#315]
I have a mapping defined...

```typescript
public mapProfile(): MappingProfile {
    return (mapper: Mapper) => {

      mapper.createMap(SectionDocumentMap, SectionDto)
      .forMember((destination) => destination.id, mapFrom((source) => source._id));
       
    };
  }

```

and I call the mapping like...

```typescript
const section = this.sectionMapper.map<SectionDocumentMap, SectionDto>(result, SectionDocumentMap, SectionDto);
```

and then get an error that the reverse of this mapping is not found

```typescript 
[Nest] 37960  - 07/16/2021, 3:51:56 PM   ERROR [AutomapperModule] Mapping not found for source class SectionDto {
} and destination class SectionDocumentMap {
}
[Nest] 38037  - 07/16/2021, 4:09:24 PM   ERROR [ExceptionsHandler] mapping is not iterable
TypeError: mapping is not iterable
```

```typescript
export class SectionDto {
  
  readonly id?: string;
  @AutoMap()
  readonly slug: string;
  @AutoMap()
  readonly name: string;
  @AutoMap()
  readonly order: number;
  @AutoMap()
  readonly published: boolean;
  @AutoMap()
  readonly description: string;
  @AutoMap()
  readonly icon: string;
  
}
```

```typescript
export class SectionDocumentMap  {
 
  _id?: string;
  @AutoMap()
  slug: string;
  @AutoMap()
  name: string;
  @AutoMap()
  order: number;
  @AutoMap()
  published: boolean;
  @AutoMap()
  description: string;
  @AutoMap()
  icon: string;

}
```

Data

```typescript
{
  _id: 60cd265fbfad853ccb0372b9,
  slug: 'housing',
  name: 'Housing',
  order: 1,
  published: true,
  description: 'this is the description of the section',
  icon: 'http://path-to-icon',
  categories: [ '60d3cdda8155eaca72b1b974' ]
}
```

**Comment by micalevisk** [comment]
They aren't required.

The error says that `SectionDto` is the source of the mapping, not the destination like you want to. You should change the order below

```js
const section = this.sectionMapper.map(result, SectionDocumentMap, SectionDto);
```
to
```js
const section = this.sectionMapper.map(result, SectionDto, SectionDocumentMap);
```

I know is somewhat weirde to read that from left to right :p

**Comment by jbeckton** [comment]
Is there a technical reason why it has to be backwards or was that a design decision? Every automapper I have ever used works the same way source -> destination.

**Comment by jbeckton** [comment]
So it looks like when you create a mapping its source -> destination and when you use it it's destination -> source?

**Comment by micalevisk** [comment]
> So it looks like when you create a mapping its source -> destination and when you use it it's destination -> source?

exactly.

> Is there a technical reason why it has to be backwards or was that a design decision?

Only @nartc could answer this one.

**Comment by nartc** [comment]
@jbeckton That is a remnant of the previous version of this library.

In the beginning, this library only works with Classes (ES6 classes). The API of `Mapper#map` was: 

```ts
Mapper.map(sourceObject, DestinationClass);
```

The API assumes `sourceObject` to be an **instance** of `SourceClass` hence it isn’t required to pass in `SourceClass`. But it wasn’t the case at all, TypeScript can be tricked easily with Type Assertion (`as SourceClass`) and the consumers could pass a Plain Object. 

Later on, to mitigate this issue, I added an optional `SourceClass` parameter to `Mapper#map`

```ts
Mapper.map(sourceObject, DestinationClass, SourceClass?)
```

When I refactored the library, I made the parameter `SourceClass` required and not have AutoMapper to try to guess anything from `sourceObject`. That’s why the API is what it is today.

I totally understand the confusion but it’s been that way for a long time now. Maybe I can make the switch when we have enough features/fixes to cut a new major version.
